tt_metal/graph_tracking.hpp

```diff
@@ -0,0 +1,113 @@
+// SPDX-FileCopyrightText: © 2023 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <string>
+#include <any>
+#include <span>
+#include <string_view>
+
+#include "tt_metal/common/core_coord.h"
+#include "tt_metal/impl/buffers/buffer.hpp"
+
+namespace tt::tt_metal {
+
+    class IGraphProcessor{
+    public:
+        IGraphProcessor() = default;
+
+        virtual void track_allocate(tt::tt_metal::Buffer* buffer, bool bottom_up) {};
+
+        virtual void track_deallocate(tt::tt_metal::Buffer* buffer) {};
+
+        virtual void track_allocate_cb(const CoreRangeSet &core_range_set, uint64_t addr, uint64_t size) {};
+
+        virtual void track_deallocate_cb() {};
+
+        virtual void track_begin_op(std::string_view function_name, std::span<std::any> input_parameters) {};
+
+        virtual void track_end_op(const std::any& output_tensors) {};
+
+
+        virtual void begin_capture(bool fake_run) {};
+        virtual std::string end_capture() {return "";};
+
+        virtual ~IGraphProcessor() = default;
+
+    };
+
+    class IGraphHooks {
+    public:
+        IGraphHooks() = default;
+        virtual bool hook_allocate(tt::tt_metal::Buffer* buffer, bool bottom_up) = 0;
+
+        virtual bool hook_deallocate(tt::tt_metal::Buffer* buffer) = 0;
+
+        virtual bool block_run_program() = 0;
```

Comment:
consider if we can really do a hook_program_run

Reply:
done